File: scripts/fetch_screener_data.py
#!/usr/bin/env python3
"""
Builds a screener data feed from a constituent list.

Runs in GitHub Actions on a daily cron. Uses yfinance (public Yahoo Finance
data) so there is no API key and no per-symbol subscription restriction -- the
whole list is refreshed every run.

Three feeds are produced by three staggered runs so the Nasdaq 100 (the
screener's default view) always refreshes first and never waits on the larger
jobs:
  --list data/nasdaq100.json --out data/screener.json          (23:00 UTC)
  --list data/sp500.json     --out data/screener_sp500.json     (23:30 UTC)
  --combined growth=data/vug.json --combined value=data/vtv.json \
      --combined dividend=data/vig.json --out data/screener_gvd.json  (00:00 UTC)
Defaults reproduce the original Nasdaq 100 behavior when run with no args.

Combined mode bundles several universes into one feed file, keyed by universe
name, with each universe holding the same {updated, source, stocks} object a
single-list feed has. Symbols appearing in more than one universe (the Growth,
Value, and Dividend lists overlap heavily) are fetched from Yahoo only once
per run.

Forward metrics use the CURRENT fiscal-year ("0y") analyst consensus to match
Seeking Alpha's "FWD" convention (not yfinance's forwardPE / "+1y" rows, which
look a year further out). P/E FWD uses Yahoo's priceEpsCurrentYear and PEG FWD
uses Yahoo's pegRatio, both of which track Seeking Alpha. EPS Growth FWD is
GAAP-basis and will differ from SA's Non-GAAP figure.

Output schema matches what screener.html reads:
  { "updated": ISO, "source": "yahoo", "stocks": { TICKER: {...}, ... } }

Env:
  PAUSE   seconds to wait between symbols (default 0.8) -- be polite to Yahoo
"""

import argparse
import datetime
import json
import os
import sys
import time

import yfinance as yf

DEFAULT_LIST = "data/nasdaq100.json"
DEFAULT_OUT = "data/screener.json"
PAUSE = float(os.environ.get("PAUSE", "0.8"))


def yahoo_symbol(sym):
    """Yahoo uses a dash for share classes that Wikipedia writes with a dot."""
    return sym.replace(".", "-")


def num(x):
    """Coerce to a finite float, else None."""
    try:
        f = float(x)
    except (TypeError, ValueError):
        return None
    if f != f or f in (float("inf"), float("-inf")):
        return None
    return f


def estimate_growth(df, period):
    """Pull the analyst 'growth' value (a decimal) for a period row, e.g. '0y'."""
    try:
        return num(df.loc[period, "growth"])
    except Exception:
        return None


def estimate_avg(df, period):
    """Pull the analyst average estimate (e.g. EPS or revenue) for a period row."""
    try:
        return num(df.loc[period, "avg"])
    except Exception:
        return None


def fetch(symbol):
    t = yf.Ticker(yahoo_symbol(symbol))
    info = t.info or {}

    rec = {}
    rec["price"] = num(info.get("currentPrice")) or num(info.get("regularMarketPrice"))
    rec["marketCap"] = num(info.get("marketCap"))
    rec["cash"] = num(info.get("totalCash"))
    rec["debt"] = num(info.get("totalDebt"))

    # Daily change: latest price vs the prior session's close. The pipeline runs
    # after the US close, so this is that trading day's move.
    prev = num(info.get("regularMarketPreviousClose")) or num(info.get("previousClose"))
    rec["prevClose"] = prev
    if rec["price"] is not None and prev not in (None, 0):
        rec["changePct"] = (rec["price"] / prev - 1) * 100
    else:
        rec["changePct"] = None

    rg = num(info.get("revenueGrowth"))
    rec["revTTM"] = rg * 100 if rg is not None else None
    eg = num(info.get("earningsGrowth"))
    rec["epsTTM"] = eg * 100 if eg is not None else None

    # Forward figures use the CURRENT fiscal-year ("0y") consensus estimate to
    # match Seeking Alpha's "FWD" convention. yfinance's forwardPE / "+1y" rows
    # are one fiscal year further out, which reads systematically too low.
    earn = None
    rev = None
    try:
        earn = t.earnings_estimate
    except Exception:
        pass
    try:
        rev = t.revenue_estimate
    except Exception:
        pass

    # P/E FWD: Yahoo's priceEpsCurrentYear (price / current-FY EPS) matches
    # Seeking Alpha's "P/E FWD" to the cent. Fall back to price / 0y estimate,
    # then to forwardPE.
    price = rec["price"]
    pe = num(info.get("priceEpsCurrentYear"))
    if pe is None:
        eps_cur = estimate_avg(earn, "0y")
        if price is not None and eps_cur is not None and eps_cur > 0:
            pe = price / eps_cur
        else:
            pe = num(info.get("forwardPE"))
    rec["peFwd"] = pe

    # Revenue / EPS Growth FWD = current-FY ("0y") consensus growth.
    # NOTE: EPS Growth FWD is GAAP-basis and will differ from Seeking Alpha,
    # which uses Non-GAAP consensus (not exposed by yfinance).
    eg_fwd = estimate_growth(earn, "0y")
    rec["epsFwd"] = eg_fwd * 100 if eg_fwd is not None else None
    rg_fwd = estimate_growth(rev, "0y")
    rec["revFwd"] = rg_fwd * 100 if rg_fwd is not None else None

    # PEG FWD: Yahoo's pegRatio incorporates a longer-term growth rate and
    # tracks Seeking Alpha's PEG (FWD) closely (SA's exact 3-5yr long-term
    # growth input is not exposed by yfinance). Fall back to trailingPegRatio,
    # then to the 1-yr forward PEG (P/E / EPS growth).
    peg = num(info.get("pegRatio"))
    if peg is None or peg == 0:
        peg = num(info.get("trailingPegRatio"))
    if (peg is None or peg == 0) and rec["peFwd"] is not None and rec["epsFwd"] not in (None, 0) and rec["epsFwd"] > 0:
        peg = rec["peFwd"] / rec["epsFwd"]
    rec["pegFwd"] = peg

    return rec


def build_stocks(listing, now, cache):
    """Fetch every symbol in a listing, reusing records already fetched this run."""
    stocks = {}
    ok = 0
    for item in listing:
        sym = item["t"]
        if sym in cache:
            rec = dict(cache[sym])
            rec["name"] = item["n"]  # each list keeps its own curated name
            stocks[sym] = rec
            if rec.get("price") is not None:
                ok += 1
            continue
        rec = {"t": sym, "name": item["n"]}
        for attempt in range(3):
            try:
                rec.update(fetch(sym))
                rec["priceUpdated"] = now
                rec["fundamentalsUpdated"] = now
                if rec.get("price") is not None:
                    ok += 1
                break
            except Exception as e:  # noqa: BLE001 - keep going on any single-symbol failure
                if attempt == 2:
                    print(f"{sym}: failed after retries: {e!r}", file=sys.stderr)
                else:
                    time.sleep(2)
        stocks[sym] = rec
        cache[sym] = rec
        time.sleep(PAUSE)
    return stocks, ok


def main():
    ap = argparse.ArgumentParser(description="Build a screener data feed from a constituent list.")
    ap.add_argument("--list", dest="list_path", default=DEFAULT_LIST,
                    help=f"constituent list JSON (default {DEFAULT_LIST})")
    ap.add_argument("--combined", action="append", default=[], metavar="NAME=LIST",
                    help="bundle several universes into one feed keyed by NAME "
                         "(repeatable, e.g. --combined growth=data/vug.json); overrides --list")
    ap.add_argument("--out", dest="out_path", default=DEFAULT_OUT,
                    help=f"output feed JSON (default {DEFAULT_OUT})")
    args = ap.parse_args()

    now = datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z")
    cache = {}  # symbol -> fetched record, shared across universes in one run

    if args.combined:
        universes = {}
        totals = []
        for spec in args.combined:
            name, _, list_path = spec.partition("=")
            if not name or not list_path:
                sys.exit(f"Bad --combined spec {spec!r}; expected NAME=LIST.")
            with open(list_path, encoding="utf-8") as f:
                listing = json.load(f)
            stocks, ok = build_stocks(listing, now, cache)
            universes[name] = {"updated": now, "source": "yahoo", "stocks": stocks}
            totals.append(f"{name} {ok}/{len(listing)}")
        out = {"updated": now, "source": "yahoo", "universes": universes}
        summary = "; ".join(totals)
    else:
        with open(args.list_path, encoding="utf-8") as f:
            listing = json.load(f)
        stocks, ok = build_stocks(listing, now, cache)
        out = {"updated": now, "source": "yahoo", "stocks": stocks}
        summary = f"{ok}/{len(listing)}"

    with open(args.out_path, "w", encoding="utf-8") as f:
        json.dump(out, f, indent=2)
        f.write("\n")

    print(f"Wrote {args.out_path}: {summary} symbols with price data.")


if __name__ == "__main__":
    main()

File: scripts/fetch-screener-data.mjs
// fetch-screener-data.mjs
//
// Builds data/screener.json for the Nasdaq 100 screener (screener.html).
// Runs in GitHub Actions on a daily cron. No npm dependencies — uses Node 20+
// global fetch and the Financial Modeling Prep "stable" API.
//
// Strategy (keeps a single daily run under the FMP free tier's 250 req/day):
//   Each run fully refreshes a rotating slice of the stalest symbols, oldest
//   first. A full refresh is 4 requests/symbol (quote, balance-sheet,
//   financial-growth, analyst-estimates). At the default of 50 symbols/run that
//   is ~200 requests, and the whole 100-name list cycles every ~2 days.
//
// Env:
//   FMP_API_KEY  (required)  - your FMP key (a GitHub Actions secret)
//   DAILY_COUNT  (optional)  - symbols to fully refresh per run (default 50)
//   REQ_BUDGET   (optional)  - hard cap on API requests this run (default 240)

import fs from "node:fs/promises";

const KEY = process.env.FMP_API_KEY;
if (!KEY) { console.error("ERROR: FMP_API_KEY is not set."); process.exit(1); }

const BASE = "https://financialmodelingprep.com/stable";
const DAILY_COUNT = parseInt(process.env.DAILY_COUNT || "50", 10);
const REQ_BUDGET = parseInt(process.env.REQ_BUDGET || "240", 10);
const LIST_PATH = "data/nasdaq100.json";
const OUT_PATH = "data/screener.json";

let reqs = 0;

function classify(body, status) {
  let msg = "";
  if (typeof body === "string") msg = body;
  else if (body && typeof body === "object" && !Array.isArray(body)) msg = body["Error Message"] || body.message || "";
  else return null; // array = success
  if (/invalid api key|legacy endpoint/i.test(msg)) return "auth";
  if (/limit reached|rate limit|too many requests/i.test(msg)) return "rate";
  if (/restricted|premium|subscription|special endpoint|not available under/i.test(msg)) return "premium";
  if (status === 401 || status === 403) return "auth";
  if (status === 429) return "rate";
  if (status === 402) return "premium";
  if (msg) return "http";
  return null;
}

async function fmp(pathq) {
  if (reqs >= REQ_BUDGET) { const e = new Error("request budget reached"); e.code = "budget"; throw e; }
  reqs++;
  const url = `${BASE}${pathq}${pathq.includes("?") ? "&" : "?"}apikey=${encodeURIComponent(KEY)}`;
  const res = await fetch(url);
  const txt = await res.text();
  let body;
  try { body = JSON.parse(txt); } catch { body = txt; }
  const code = classify(body, res.status);
  if (code) {
    const err = new Error(typeof body === "string" ? body : (body && body["Error Message"]) || ("HTTP " + res.status));
    err.code = code;
    throw err;
  }
  return body;
}

const isNum = (n) => typeof n === "number" && isFinite(n);

function nearestForwardPair(est) {
  if (!Array.isArray(est) || !est.length) return null;
  const s = [...est].sort((a, b) => new Date(a.date) - new Date(b.date));
  const now = Date.now();
  let idx = s.findIndex((e) => new Date(e.date).getTime() >= now);
  if (idx === -1) idx = s.length - 1;
  let fy0 = s[idx], fy1 = s[idx + 1] || null;
  if (!fy1 && idx > 0) { fy1 = fy0; fy0 = s[idx - 1]; }
  return { fy0, fy1 };
}

function recomputeValuation(rec) {
  if (isNum(rec.price) && isNum(rec.fwdEps) && rec.fwdEps > 0) {
    rec.peFwd = rec.price / rec.fwdEps;
    rec.pegFwd = isNum(rec.epsFwd) && rec.epsFwd > 0 ? rec.peFwd / rec.epsFwd : null;
  }
}

async function updatePrice(rec) {
  const q = await fmp(`/quote?symbol=${rec.t}`);
  if (q && q[0]) {
    if (isNum(q[0].price)) rec.price = q[0].price;
    if (isNum(q[0].marketCap)) rec.marketCap = q[0].marketCap;
  }
  recomputeValuation(rec);
  rec.priceUpdated = new Date().toISOString();
}

async function updateFundamentals(rec) {
  const bs = await fmp(`/balance-sheet-statement?symbol=${rec.t}&period=annual&limit=1`);
  if (bs && bs[0]) {
    rec.cash = isNum(bs[0].cashAndShortTermInvestments) ? bs[0].cashAndShortTermInvestments : null;
    rec.debt = isNum(bs[0].totalDebt) ? bs[0].totalDebt : null;
  }
  const fg = await fmp(`/financial-growth?symbol=${rec.t}&period=annual&limit=1`);
  if (fg && fg[0]) {
    rec.revTTM = isNum(fg[0].revenueGrowth) ? fg[0].revenueGrowth * 100 : null;
    rec.epsTTM = isNum(fg[0].epsgrowth) ? fg[0].epsgrowth * 100 : null;
  }
  const est = await fmp(`/analyst-estimates?symbol=${rec.t}&period=annual&limit=8`);
  const pair = nearestForwardPair(est);
  if (pair && pair.fy0) {
    const e0 = pair.fy0, e1 = pair.fy1;
    rec.fwdEps = isNum(e0.epsAvg) ? e0.epsAvg : null;
    if (e1 && isNum(e0.epsAvg) && e0.epsAvg !== 0 && isNum(e1.epsAvg)) rec.epsFwd = (e1.epsAvg / Math.abs(e0.epsAvg) - 1) * 100;
    if (e1 && isNum(e0.revenueAvg) && e0.revenueAvg > 0 && isNum(e1.revenueAvg)) rec.revFwd = (e1.revenueAvg / e0.revenueAvg - 1) * 100;
  }
  recomputeValuation(rec);
  rec.fundamentalsUpdated = new Date().toISOString();
}

async function main() {
  const list = JSON.parse(await fs.readFile(LIST_PATH, "utf8"));

  let prev = { stocks: {} };
  try { prev = JSON.parse(await fs.readFile(OUT_PATH, "utf8")); } catch { /* first run */ }

  const stocks = {};
  for (const item of list) {
    const t = item.t;
    stocks[t] = Object.assign({ t: t, name: item.n }, prev.stocks[t] || {});
    stocks[t].t = t; stocks[t].name = item.n;
  }

  let stopped = null;

  // Order by the stalest of a symbol's two timestamps (0 = never fetched), so
  // each run fully refreshes the least-recently-updated names first.
  const stalestMs = (t) => {
    const s = stocks[t];
    const a = s.priceUpdated ? Date.parse(s.priceUpdated) : 0;
    const b = s.fundamentalsUpdated ? Date.parse(s.fundamentalsUpdated) : 0;
    return Math.min(a, b);
  };
  const slice = list.map((i) => i.t).sort((a, b) => stalestMs(a) - stalestMs(b)).slice(0, DAILY_COUNT);

  let refreshed = 0;
  for (const t of slice) {
    if (stopped) break;
    try {
      await updatePrice(stocks[t]);
      await updateFundamentals(stocks[t]);
      refreshed++;
    } catch (e) {
      if (e.code === "auth" || e.code === "rate" || e.code === "budget") { stopped = e; break; }
      console.warn(`${t}: ${e.code || ""} ${e.message}`);
    }
  }

  const out = { updated: new Date().toISOString(), source: "fmp", requests: reqs, stocks: stocks };
  await fs.writeFile(OUT_PATH, JSON.stringify(out, null, 2) + "\n");

  const withFund = Object.values(stocks).filter((s) => s.fundamentalsUpdated).length;
  console.log(`Wrote ${OUT_PATH}: refreshed ${refreshed}/${slice.length} this run, ${withFund}/${Object.keys(stocks).length} have fundamentals, ${reqs} API requests.`);
  if (stopped) {
    console.log(`Stopped early (${stopped.code}: ${stopped.message}). Progress saved; the next run resumes.`);
    if (stopped.code === "auth") process.exit(1); // surface bad key as a failed job
  }
}

main().catch((e) => { console.error(e); process.exit(1); });
